### PR TITLE
multiplex storage to allow for choosing database backend

### PR DESCRIPTION
**Pull request by sbhatnagar30**
This will allow storage to use Mongo to store the wallet structure

Also allows wallet lite functionality for a read only wallet structure

### CODE REVIEW

File: packages/bitcore-client/tsconfig.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
   },
   "include": [
     "./src/**/*.ts",
+    "./src/storage/**/*.ts",
```
Comment (micahriggan) [inline packages/bitcore-client/tsconfig.json]: this shouldn't need to be there

File: packages/bitcore-client/src/storage/mongo.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+export class Mongo {
+  path: string;
+  db: any;
+  collectionName: string;
+  collection: any;
+  errorIfExists?: boolean;
+  createIfMissing: boolean;
+  dbName: string;
+  databaseName: string;
+  client: any;
```
Comment (micahriggan) [inline packages/bitcore-client/src/storage/mongo.ts]: Mongo.Client

File: packages/bitcore-client/src/storage/mongo.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
+import 'source-map-support/register'
+import { MongoClient } from 'mongodb'
+import {baToJSON} from "ethereumjs-util";
+import {Transform} from "stream";
+
+export class Mongo {
+  path: string;
+  db: any;
```
Comment (micahriggan) [inline packages/bitcore-client/src/storage/mongo.ts]: Mongo.Db

File: packages/bitcore-client/src/storage/mongo.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  async listWallets() {
+    await this.init({wallet: 1});
+    const stream = new Transform({
+      objectMode: true,
+      transform(data, enc, next) {
+        this.push(JSON.stringify(data));
+        next();
+      }
+    });
+    const cursor = this.collection.find({'name': { '$exists': true }}, { 'name': 1, 'chain': 1, 'network': 1 }).pipe(stream);
```
Comment (micahriggan) [inline packages/bitcore-client/src/storage/mongo.ts]: likely want to index on name

something like createIndex({name: 1}) should probably be in init

File: packages/bitcore-client/bin/wallet-create-lite [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env node
```
Comment (micahriggan) [inline packages/bitcore-client/bin/wallet-create-lite]: curious why a separate script over adding the --lite flag

Comment (sbhatnagar30) [reply packages/bitcore-client/bin/wallet-create-lite]: I could do that, I don't think there was a particular reason

File: packages/bitcore-client/src/storage.ts [hunk 6]
```diff
@@ -1,116 +1,77 @@
...
-    } else {
-      console.log('using wallets at', this.path);
-      this.db = StorageCache[this.path] = levelup(lvldwn(this.path), {
-        createIfMissing,
-        errorIfExists
-      });
+    this.path = path;
+    this.createIfMissing = createIfMissing;
+    this.errorIfExists = errorIfExists;
+    this.dbName = dbName;
```
Comment (micahriggan) [inline packages/bitcore-client/src/storage.ts]: could your refactor the name of this to "storageType"

File: packages/bitcore-client/bin/wallet-create [hunk 7]
```diff
@@ -28,7 +33,11 @@ const main = async () => {
       network,
       path,
       phrase,
-      password
+      password,
```
Comment (micahriggan) [inline packages/bitcore-client/bin/wallet-create]: You refactored to storageType right?

Comment (sbhatnagar30) [reply packages/bitcore-client/bin/wallet-create]: missed some :sweat_smile:

File: packages/bitcore-client/src/storage.ts [hunk 8]
```diff
@@ -1,106 +1,99 @@
...
+      listWalletStream.once('end', () => --this.db.length === 0 && passThrough.end());
+    }
+    return passThrough;
   }
 
   async saveWallet(params) {
     const { wallet } = params;
-    return this.db.put(`wallet|${wallet.name}`, JSON.stringify(wallet));
+    return this.storageType.saveWallet({ wallet });
+    // return this.db.find(db => db === this.storageType).saveWallet({ wallet });
```
Comment (micahriggan) [inline packages/bitcore-client/src/storage.ts]: remove commented line

File: packages/bitcore-client/src/storage.ts [hunk 9]
```diff
@@ -1,106 +1,99 @@
...
-    const payload = (await this.db.get(`key|${name}|${address}`)) as string;
+  async getKey(params: {
+    address: string;
+    name: string;
+    encryptionKey: string;
+    keepAlive: boolean;
+    open: boolean;
+  }): Promise<KeyImport> {
+    const { address, name, encryptionKey, keepAlive, open } = params;
+    // const payload = await this.db.find(db => db === this.storageType).getKey({ name, address, keepAlive, open });
```
Comment (micahriggan) [inline packages/bitcore-client/src/storage.ts]: remove commented line

File: packages/bitcore-client/src/storage.ts [hunk 10]
```diff
@@ -137,7 +143,13 @@ export class Storage {
         payload = { encKey, pubKey };
       }
       const toStore = JSON.stringify(payload);
-      await this.db.put(`key|${name}|${key.address}`, toStore);
+      let keepAlive = true;
+      if (key === keys[keys.length - 1]) {
+        keepAlive = false;
+      }
+      await this.storageType.addKeys({ name, key, toStore, keepAlive, open });
+      // await this.db.find(db => db == this.storageType).addKeys({ name, key, toStore, keepAlive, open });
```
Comment (micahriggan) [inline packages/bitcore-client/src/storage.ts]: remove commented line

File: packages/bitcore-client/src/wallet.ts [hunk 11]
```diff
@@ -91,22 +93,24 @@ export class Wallet {
...
     let alreadyExists;
     try {
-      alreadyExists = await this.loadWallet({ storage, name });
+      alreadyExists = await this.loadWallet({ storage, name, storageType });
     } catch (err) {}
     if (alreadyExists) {
       throw new Error('Wallet already exists');
     }
-
+    console.log('Creating new wallet...');
```
Comment (micahriggan) [inline packages/bitcore-client/src/wallet.ts]: remove console log

File: packages/bitcore-client/src/storage.ts [hunk 12]
```diff
@@ -1,106 +1,99 @@
...
+    if (dbMap[storageType]) {
+      this.db.push(new dbMap[storageType]({ createIfMissing, errorIfExists, path }));
+      this.storageType = this.db[0];
     } else {
-      console.log('using wallets at', this.path);
-      this.db = StorageCache[this.path] = levelup(lvldwn(this.path), {
-        createIfMissing,
-        errorIfExists
-      });
+      for (let dbMapKey in dbMap) {
```
Comment (micahriggan) [inline packages/bitcore-client/src/storage.ts]: can you change it to this
```
for (let dbType of Object.values(dbMap))
```

File: packages/bitcore-client/src/wallet.ts [hunk 13]
```diff
@@ -149,12 +155,17 @@ export class Wallet {
...
     let { storage } = params;
-    storage = storage || new Storage({ errorIfExists: false, createIfMissing: false, path });
+    storage = storage || new Storage({ errorIfExists: false, createIfMissing: false, path, storageType });
     const loadedWallet = await storage.loadWallet({ name });
-    return new Wallet(Object.assign(loadedWallet, { storage }));
+    try {
+      return new Wallet(Object.assign(loadedWallet, { storage }));
+    } catch (err) {
+      console.error('Could not find wallet');
+      throw new Error(err);
```
Comment (micahriggan) [inline packages/bitcore-client/src/wallet.ts]: This feels strange to me. Getting an error "Could not find wallet" that is triggered from the wallet constructor failing, but not from storage.loadWallet..

And then the catch just re-throws, can we remove it?

Seems weird.

File: packages/bitcore-node/test/integration/wallet-benchmark.integration.ts [hunk 14]
```diff
@@ -282,13 +282,13 @@ describe('Wallet Benchmark', function() {
       expect(largeAddressBatch.length).to.deep.equal(1000);
 
       console.log('Checking');
-      const importedWallet1 = await createWallet(smallAddressBatch, 0, network);
-      const importedWallet2 = await createWallet(mediumAddressBatch, 1, network);
-      const importedWallet3 = await createWallet(largeAddressBatch, 2, network);
+      const importedWallet4 = await createWallet(smallAddressBatch, 4, network);
```
Comment (micahriggan) [inline packages/bitcore-node/test/integration/wallet-benchmark.integration.ts]: These shouldn't need to be changed, createWallet loads the wallet or creates it if it doesn't exist. This should just load the previously created wallet.

**Comment by matiu** [comment]
For a future PR, passing `storageType` to all command could be simplified by using a ENV variable or a global config.

**Pull request merged**